src/modules/bulletproofs/main_impl.h

```diff
@@ -0,0 +1,237 @@
+ const uint64_t *min_value, const secp256k1_pedersen_commitment* commit, size_t n_commits, size_t nbits, const secp256k1_generator *value_gen, const unsigned char *extra_commit, size_t extra_commit_len) {
+    int ret;
+    size_t i;
+    secp256k1_ge *commitp;
+    secp256k1_ge value_genp;
+    const secp256k1_ge *commitp_ptr;
+    const uint64_t *minvalue_ptr;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(scratch != NULL);
+    ARG_CHECK(gens != NULL);
+    ARG_CHECK(gens->n >= 2 * nbits * n_commits);
+    ARG_CHECK(proof != NULL);
+    ARG_CHECK(commit != NULL);
+    ARG_CHECK(n_commits > 0);
+    ARG_CHECK(nbits > 0);
+    ARG_CHECK(nbits <= 64);
+    ARG_CHECK(value_gen != NULL);
+    ARG_CHECK(extra_commit != NULL || extra_commit_len == 0);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+
+    if (!secp256k1_scratch_allocate_frame(scratch, 2 * n_commits * sizeof(secp256k1_ge), 1)) {
+        return 0;
+    }
+
+    commitp = (secp256k1_ge *)secp256k1_scratch_alloc(scratch, n_commits * sizeof(secp256k1_ge));
+    for (i = 0; i < n_commits; i++) {
+        secp256k1_pedersen_commitment_load(&commitp[i], &commit[i]);
+    }
+    secp256k1_generator_load(&value_genp, value_gen);
+
+    commitp_ptr = commitp;
+    minvalue_ptr = min_value;
+    ret = secp256k1_bulletproof_rangeproof_verify_impl(&ctx->ecmult_ctx, scratch, &proof, 1, plen, nbits, &minvalue_ptr, &commitp_ptr, n_commits, &value_genp, gens, &extra_commit, &extra_commit_len);
+    secp256k1_scratch_deallocate_frame(scratch);
+    return ret;
+}
+
+int secp256k1_bulletproof_rangeproof_verify_multi(const secp256k1_context* ctx, secp256k1_scratch_space *scratch, const secp256k1_bulletproof_generators *gens, const unsigned char* const* proof, size_t n_proofs, size_t plen, const uint64_t* const* min_value, const secp256k1_pedersen_commitment* const* commit, size_t n_commits, size_t nbits, const secp256k1_generator *value_gen, const unsigned char* const* extra_commit, size_t *extra_commit_len) {
+    int ret;
+    secp256k1_ge **commitp;
+    secp256k1_ge *value_genp;
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(scratch != NULL);
+    ARG_CHECK(gens != NULL);
+    ARG_CHECK(gens->n >= 2 * nbits * n_commits);
+    ARG_CHECK(commit != NULL);
+    ARG_CHECK(proof != NULL);
+    ARG_CHECK(n_proofs > 0);
+    ARG_CHECK(n_commits > 0);
+    ARG_CHECK(nbits > 0);
+    ARG_CHECK(nbits <= 64);
+    ARG_CHECK(value_gen != NULL);
+    ARG_CHECK((extra_commit_len == NULL) == (extra_commit == NULL));
+    if (extra_commit != NULL) {
+        for (i = 0; i < n_proofs; i++) {
+            ARG_CHECK(extra_commit[i] != NULL || extra_commit_len[i] == 0);
+        }
+    }
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+
+    if (!secp256k1_scratch_allocate_frame(scratch, n_proofs * (sizeof(*value_genp) + sizeof(*commitp) + n_commits * sizeof(**commitp)), 1 + n_proofs)) {
```

Comment:
According to @jasonxh 's finding in https://github.com/mimblewimble/secp256k1-zkp/pull/38,  here the `1 + n_proofs)` should be `2 + n_proofs)`, since we have `secp256k1_scratch_alloc` at line 148, 149 and `n_proofs` loops at line 150.

Reply:
Great catch, thanks!

Reply:
Fixed.